byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,404 @@
+package byzcoin
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"go.dedis.ch/cothority/v3/darc"
+	"go.dedis.ch/protobuf"
+)
+
+// The deferred contract allows a group of signers to agree on and sign a
+// proposed transaction, the "proposed transaction".
+
+// ContractDeferredID denotes a contract that can aggregate signatures for a
+// "proposed" transaction
+var ContractDeferredID = "deferred"
+
+const defaultNumExecution uint64 = 1
+
+// DeferredData contains the specific data of a deferred contract
+type DeferredData struct {
+	// The transaction that signers must sign and can be executed with an
+	// "executeProposedTx".
+	ProposedTransaction ClientTransaction
+	// If the current block index is greater than this value, any Invoke on the
+	// deferred contract is rejected. This provides an expiration mechanism.
+	ExpireBlockIndex uint64
+	// Hashes of each instruction of the proposed transaction. Those hashes are
+	// computed using the special "hashDeferred" method.
+	InstructionHashes [][]byte
+	// The number of time the proposed transaction can be executed. This number
+	// decreases for each successful invocation of "executeProposedTx" and its
+	// default value is set to 1.
+	NumExecution uint64
+	// This array is filled with the instruction IDs of each executed
+	// instruction when a successful "executeProposedTx" happens.
+	ExecResult [][]byte
+}
+
+type contractDeferred struct {
+	BasicContract
+	DeferredData
+	s *Service
+}
+
+func (s *Service) contractDeferredFromBytes(in []byte) (Contract, error) {
+	c := &contractDeferred{s: s}
+
+	err := protobuf.Decode(in, &c.DeferredData)
+	if err != nil {
+		return nil, errors.New("couldn't unmarshal instance data: " + err.Error())
+	}
+	return c, nil
+}
+
+func (c *contractDeferred) Spawn(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
+	// This method should do the following:
+	//   1. Parse the input buffer
+	//   2. Compute and store the instruction hashes
+	//   3. Save the data
+	//
+	// Spawn should have those input arguments:
+	//   - proposedTransaction ClientTransaction
+	//   - expireBlockIndex uint64
+	//   - numExecution uint64 (default: 1)
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	// 1. Reads and parses the input
+	proposedTransaction := ClientTransaction{}
+	err = protobuf.Decode(inst.Spawn.Args.Search("proposedTransaction"), &proposedTransaction)
+	if err != nil {
+		return nil, nil, errors.New("couldn't decode proposedTransaction: " + err.Error())
+	}
+	expireBlockIndex, err := strconv.ParseUint(string(inst.Spawn.Args.Search("expireBlockIndex")), 10, 64)
```

Comment:
integers should be serialized as the byte representation of the integer and not the byte representation of the string representation of the integer. in our code we usually use `binary.LittleEndian.PutUint64`